### ISSUE TITLE
Do not compare selectors ending in pseudo classes to selectors ending in pseudo elements in no-descending-specificity

### ISSUE THREAD

**Issue by eviltwin** [stylelint/stylelint#1811]
With the following CSS:

```css
.Button {
  display: inline-block;
  margin: 0;
  padding: 0.75em 1em;
}

.Button:not(.disabled):active {
  padding-top: calc(0.75em + 1px);
  padding-bottom: calc(0.75em - 1px);
  box-shadow: inset 0 1px 0 rgba(0, 0, 0, 0.16);
}

.Button::-moz-focus-inner {
  border: 0;
  padding: 0;
}
```

and the following `.stylelintrc` in the same directory:

```json
{
  "rules": {
    "no-descending-specificity": true
  }
}
```

When I run the command: `stylelint Button.css`

I get the warning:

```shell
button.css
 13:1  ✖  Expected selector ".button::-moz-focus-inner" to come before selector ".button:not(.disabled):active"   no-descending-specificity
```

but, since `.Button::-moz-focus-inner` is actually selecting a pseudo element within the `.Button`, the rules don't conflict and I wouldn't expect to get a warning about specificity. (Unless my understanding of the CSS is flawed here..?)

stylelint version 7.1.0

**Comment by jeddy3** [comment]
The [purpose](https://github.com/stylelint/stylelint/tree/master/src/rules/no-descending-specificity#no-descending-specificity) of this rule is to remove confusion, rather than conflict, by *aligning the source order and specificity*. 

As `.Button:not(.disabled):active` has a specificity of `0,3,0` and `.Button::-moz-focus-inner` of `0,1,1` it makes sense that this rule recommends having `.Button::-moz-focus-inner` come before `.Button:not(.disabled):active`. That way the source order and the specificity are aligned.

Feel free to reopen this issue if this doesn't address the issue.

**Comment by eviltwin** [comment]
from the rule description:

> Here's how it works: **This rule looks at the last compound selector in every full selector, and then compares it with other selectors in the stylesheet that end in the same way.**
>
> So `.foo .bar` (whose last compound selector is `.bar`) will be compared to `.bar` and `#baz .bar`, but not to `#baz .foo` or `.bar .foo`.

To me, this strongly implies that the purpose of the rule is to align the source order and specificity of rules _that are likely to target the same element_. I'd argue that the last compound selector of `.Button::-moz-focus-inner` is actually `-moz-focus-inner` since the pseudo-element selector is targeting a child of `.Button`.

**Comment by eviltwin** [comment]
(also it seems that I don't have the appropriate permissions to reopen this issue, so could you please do so if my argument is persuasive enough?)

**Comment by davidtheclark** [comment]
@eviltwin I think your argument here would be better illustrated with a PR showing some test cases.

**Comment by davidtheclark** [comment]
On thinking about this a little I believe I'm probably in agreement with @eviltwin. Maybe we should make the compound selector include any pseudo-elements in the selector. 

I think that this would make the rule *more lenient* so would not require a major version bump. That's nice.

I will label this issue as help wanted. @eviltwin if you'd like to see this feature soon, please consider contributing a PR to change the rule's behavior.

Does anybody see a problem with this reasoning?

**Comment by jeddy3** [comment]
> Maybe we should make the compound selector include any pseudo-elements in the selector.

Sgtm.

Pseudo-elements [seem to be a bit special](https://drafts.csswg.org/selectors-4/#pseudo-element-attachment). I think treating them in the way your described makes sense.
